File: 0124-binary-tree-maximum-path-sum/0124-binary-tree-maximum-path-sum.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int ans = -1001;
    int maxPathSum(TreeNode* root)
    {
        dfs(root);
        return ans;
    }
    
    int dfs(TreeNode* root)
    {
        if(root == nullptr) return 0;
        
        int left = dfs(root->left);
        int right = dfs(root->right);
        
        int curt = root->val;
        
        if(left > 0) curt += left;
        if(right > 0) curt += right;
        
        ans = max(ans, curt);
        
        return max(root->val, max(root->val + left, root->val + right));
    }
};

File: 0124-binary-tree-maximum-path-sum/0124-binary-tree-maximum-path-sum.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def maxPathSum(self, root: Optional[TreeNode]) -> int:
        ret = [-1000]
        
        def dfs(root):
            if not root:
                return 0
            
            left = dfs(root.left)
            right = dfs(root.right)
            
            
            curt = root.val
            
            if left > 0:
                curt += left
            
            if right > 0:
                curt += right
                
            ret[0] = max(ret[0], root.val, root.val + left, root.val + right, curt)
            return max(root.val, root.val + left, root.val + right)
        dfs(root)
        return ret[0]